src/internal/connector/exchange/service_restore.go

```diff
@@ -651,19 +651,27 @@ func establishContactsRestoreLocation(
 
 	ctx = clues.Add(ctx, "is_new_cache", isNewCache)
 
-	temp, err := ac.Contacts().CreateContactFolder(ctx, user, folders[0])
-	if err != nil {
+	fold, err := ac.Contacts().CreateContactFolder(ctx, user, folders[0])
+	// 409 handling: Fetch folder if it exists.
+	// This is rare, but it may happen if an earlier CreateContactFolder POST failed
+	// with 5xx, potentially leaving dirty state in graph.
+	if graph.IsErrFolderExists(err) {
+		fold, err = ac.Contacts().GetContactFolderByDisplayName(ctx, user, folders[0])
+		if err != nil {
+			return "", err
```

Comment:
The inner error here kinda slices the outer one.  A little extra context will go a long way when trying to discover what processing occurred.
```suggestion
			return "", clues.Wrap(err, "getting folder after collision on create")
```

Reply:
100%